composeApp/src/commonMain/kotlin/sp/bvantur/tasky/register/presentation/RegisterViewModel.kt

```diff
@@ -0,0 +1,120 @@
+package sp.bvantur.tasky.register.presentation
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import sp.bvantur.tasky.core.DispatcherProvider
+import sp.bvantur.tasky.core.domain.ValidateEmailUseCase
+import sp.bvantur.tasky.core.domain.ValidateNameUseCase
+import sp.bvantur.tasky.core.domain.ValidatePasswordUseCase
+import sp.bvantur.tasky.core.presentation.ViewModelUserActionHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandlerImpl
+import kotlin.time.Duration.Companion.milliseconds
+
+typealias OnRegisterUserAction = (RegisterUserAction) -> Unit
+
+@OptIn(FlowPreview::class)
+class RegisterViewModel(
+    private val dispatcherProvider: DispatcherProvider,
+    private val validateNameUseCase: ValidateNameUseCase,
+    private val validateEmailUseCase: ValidateEmailUseCase,
+    private val validatePasswordUseCase: ValidatePasswordUseCase
+) : ViewModel(),
+    ViewModelUserActionHandler<RegisterUserAction>,
+    ViewModelViewStateHandler<RegisterViewState> by ViewModelViewStateHandlerImpl(
+        RegisterViewState(),
+        dispatcherProvider
+    ) {
+
+    private val textDebounceFlow = MutableSharedFlow<Pair<String, suspend (String) -> Unit>>()
+
+    init {
+        textDebounceFlow
+            .debounce(500.milliseconds)
+            .onEach { (text, action) ->
+                action(text)
+            }
+            .launchIn(CoroutineScope(dispatcherProvider.default))
+    }
```

Comment:
Thanks for the learning resources. I will take a look. I've also removed this part of the code from the project.

Reply:
@philipplackner & @Kai-Cilliers I've tried out this approach and have a question for you two. I see a good use case for the mentioned approach in the video and in articles where you load the data from the database and it is not that expensive to do it. But if you want to have only backend communication with this approach, from my observation is not the best fit, or maybe I implemented something wrong. 

SCENARIO
Let's say you have a first screen that loads data from the backend and presents them in the list. Then click on a single item in the list and you are navigated to a new screen. And then you stay on that new screen for more than 5 seconds. When you come back to the list the data loads again, because there are no more active subscribers. If the list has static data and only the backend is a source of data for the whole app, then the init block of ViewModel would be the better usage for that, or am I missing something here?

This is my code for that:
```
abstract class ViewStateViewModel<State : ViewState>(
    initialViewState: State,
    private val dispatcherProvider: DispatcherProvider
) : ViewModel() {

    private val mutableViewStateFlow = MutableStateFlow(initialViewState)
    val viewStateFlow: StateFlow<State> = mutableViewStateFlow.onStart { initialLoadData() }
        .stateIn(
            viewModelScope,
            SharingStarted.WhileSubscribed(SUBSCRIPTION_TIMEOUT),
            initialViewState
        )

    suspend fun emitViewState(onUpdateViewState: (State) -> State) {
        withContext(dispatcherProvider.main.immediate) {
            mutableViewStateFlow.update {
                onUpdateViewState(viewStateFlow.value)
            }
        }
    }

    protected open fun initialLoadData() {}

    private companion object {
        const val SUBSCRIPTION_TIMEOUT = 5000L
    }
}
```

Reply:
You have valid concerns which were also my concerns. I also had a few discussions about this case with different people and you have two options to handle this:
1. Once data was loaded, you save it in a separate variable. Before loading initial data, you simply then check if it was loaded before and if so, don't do that again. 
2. You could use `SharingStarted.Lazily` which will kick off once the Flow is subscribed, but then keep the result. The disadvantage is that the Flow will then remain launched when the app is in the background (which for a simple Flow like here shouldn't be a big deal, but it would if it had more intermediate operators that would keep on executing)

Reply:
I am solving the assignment for one job opportunity and I don't want to complicate with database layer since is not required to be written. Now I need to think about what would be the best approach for my use case. Thank you for your input.